### DISCUSSION TITLE
What to do about some 19318 warnings in Eclipse?

### DISCUSSION THREAD

**Discussion by gnl42** [eclipse-mylyn/org.eclipse.mylyn#409] [General]
Eclipse is currently showing (for me) 19318 warnings for Mylyn:
Of those, the major offenders are:

- 'Non-externalized string': 15478
- 'Discouraged access': 2315
- 'deprecated': 839

Any thoughts on how clean up the code?
@merks @BeckerFrank @akurtakov @ruspl-afed

**Comment by ruspl-afed** [comment]
I'm not sure what do you mean by "how". By fixing them all, of cause.

> * 'Non-externalized string'

We can suppress them for test classes, all the rest needs individual review, followed either by i18n or `//NON-NLS`

> * 'Discouraged access':

We can suppress them for test classes and may be specify `x-friends` inside one subsystem. But all the rest should be reviewed.
And I will veto blind promotion to API for packages, even if they looks like API and were used for years as an API. Because at this level of project resources, we will not be able to support old-fashion signatures promoted to the API.

> 'deprecated'

We need to review and rework them individually

**Reply by merks** [reply]
FYI, `@SuppressWarnings("nls")` works on larger scopes where appropriate, i.e., in a method or class where externalization does not and will never make sense.

**Reply by ruspl-afed** [reply]
Yes, I know, missed this in my comment for some reason. Thank you for this useful addition.

**Comment by gnl42** [comment]
"how" as in how to tackle the problem given the numbers

**Reply by merks** [reply]
Work very hard and very long.  😱 

Note that access rules can also be used to fix access warnings:

![image](https://github.com/eclipse-mylyn/org.eclipse.mylyn/assets/208716/ec394526-21e6-4751-9672-e0a28a25d65a)

For those you can fix many things quickly with small changes.

There are quick fixes for the NLS things, including the externalization wizard, but I'd start by identifying those places, e.g., tests where it just isn't needed.

I think no matter way, it's a lot of work...

**Reply by ruspl-afed** [reply]
Yes, a lot of work, but even after first approach to "commons" component I have 18096 warnings left. So, not so hopeless.

**Reply by ruspl-afed** [reply]
After completing the first round we have <7k (instead of >19k in the beginning) and there are still items relatively easy to resolve.
Thanks @gnl42 for your outstanding contribution to this activity!
Let's continue after [4.2.0](https://github.com/eclipse-mylyn/org.eclipse.mylyn/milestone/5)

**Comment by gnl42** [comment]
Re 'Discouraged access':
@ruspl-afed @merks How about for now we changed the notification level from `Waring `to `Info `for all projects.
That way they won't clutter up the things we can fix, but they are still reported

**Reply by ruspl-afed** [reply]
`Discouraged access` is not different from other kinds of warnings and indicates that we need to pay attention on it. And we can apply bulk solution in 2 base cases:
1. we are in test bundle and access non-API signatures from bundle-under-test: 
      => **suppress them** (actually this is weak solution since it may affect other usages)
2. we access non-API signatures from the same subsystem, i.e. `x.y.z.ui` calls `x.y.z.internal.core`: 
      => **declare `x-friend` `x.y.z.internal.core` package for `x.y.z.ui`**

Other cases require individual review and may be even individual tickets, because they indicates "lack of API", or "leak of API" or other things to be considered, that may lead to **[breaking] API change(s)**.

**Comment by gnl42** [comment]
I'm just suggesting that we set the level to Info temporarily so that the Warnings don't hide the items we can fix more easily

**Reply by ruspl-afed** [reply]
Eclipse Problem view has an excellent filtering capabilities to let one focus on issues of particular origin

![image](https://github.com/eclipse-mylyn/org.eclipse.mylyn/assets/15242312/00ed5811-4ffe-45ae-8e0b-bc64c98611d5)

**Comment by gnl42** [comment]
Down to 2638 (1749 'Discouraged access')
